keps/74-support-argo-workflow/README.md

```diff
@@ -0,0 +1,350 @@
+# KEP-74: Support Argo Workflow
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+- [Design Details](#design-details)
+  - [Workflow as An Unit](#workflow-as-an-unit)
+    - [Drawback and Limitations](#drawback-and-limitations)
+    - [Advantages](#advantages)
+  - [Layer as An Unit](#layer-as-an-unit)
+    - [Examples](#examples)
+      - [Example 1 (ParallelSteps Contains Leaf Template Only)](#example-1-parallelsteps-contains-leaf-template-only)
+      - [Example 2 (ParallelSteps Contains Leaf Template and Step)](#example-2-parallelsteps-contains-leaf-template-and-step)
+      - [Example 3 (Workflow with Single Container Template)](#example-3-workflow-with-single-container-template)
+    - [How to suspend a workflow step by step](#how-to-suspend-a-workflow-step-by-step)
+    - [Drawback and Limitations](#drawback-and-limitations-1)
+    - [Advantages](#advantages-1)
+  - [Plain Pod as An Unit](#plain-pod-as-an-unit)
+    - [Drawback and Limitations](#drawback-and-limitations-2)
+    - [Advantages](#advantages-2)
+- [Additional Details](#additional-details)
+  - [Test Plan](#test-plan)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This KEP outlines the proposal to integrate Argo Workflows within Kueue, discussing the advantages 
```

Comment:
I think I'd like to see the proposal speaking generically about adding support for workflow logic, with Argo as an example. We should be able to think about this design via any workflow tool that can conform to the communications defined with kueue.

Reply:
I've been thinking about this use case too, and I think what would be interesting to try is a general strategy to submit the entire DAG, and then use suspend (on the level of the job) to control when steps are actually run, and have the workload manager checking status and un-suspend (or cancel) based on the that. The important detail here is we aren't adding additional burden on kueue - it works as expected, and the bulk or orchestration is handled by the workflow tool.

In terms of "which queue" the simplest first design is for the workflow manager to decide that. There could be more complex logic after that, but then you venture into multi-cluster space and that might be beyond the discussion in this issue. A few questions I'm still thinking about (and I'll share with others here):

1. What kind of additional load are we placing on APIs for kueue/kubernetes in having as workflow tool that is going to be regularly checking state to unsuspend jobs? Is it reasonable? If yes, ok. If not, what other means can we offer to check updates of state that won't add that burden?
2. Arguably, we could use jobs with suspend already without requiring kueue. What essential feature is kueue adding, and what is the simplest way to achieve it without (so to speak) hard coding workflow manager or workflow specific tasks into kueue?
3. This might be more general to kubernetes, but I haven't found a good answer to how to handle steps that require shared inputs / outputs. Some tools handle retrieval / download between steps, and I'm not sure what people are doing these days using kueue. Is this an issue that has come up, and what are people doing?

Storage and sharing data between workflow steps, especially if the clusters are separate, seems like a hard problem. What does Argo do @KunWuLuan ?

Reply:
And I wonder if kueue could have a developer API, something with grpc, that would allow easier integration with workflow tools that can interact with that interface.

Reply:
> Storage and sharing data between workflow steps, especially if the clusters are separate, seems like a hard problem.

Hi, @terrytangyuan could you help to answer the question? Thanks very much.

Reply:
> Arguably, we could use jobs with suspend already without requiring kueue. What essential feature is kueue adding

I think what kueue do is exactly to decide when and where to place the workload.

> and what is the simplest way to achieve it without (so to speak) hard coding workflow manager or workflow specific tasks into kueue?

If all workflow managers support a standard API or same field path to indicate the specific stage or workflow should run after with admission, the goal that a generic workflow controller in Kueue can be achieved, just like what training-operator did.

Reply:
> I think I'd like to see the proposal speaking generically about adding support for workflow logic, with Argo as an example. We should be able to think about this design via any workflow tool that can conform to the communications defined with kueue.

Good idea, thanks! I have update the kep according to the suggestion.